resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala

```diff
@@ -0,0 +1,179 @@
+   }
+   private val newHadoopConfigMapName =
+     if (hadoopConfDirSpec.hadoopConfigMapName.isEmpty) {
+        Some(kubernetesConf.hadoopConfigMapName)
+     } else {
+        None
+     }
+
+   // Either use pre-existing secret or login to create new Secret with DT stored within
+   private val kerberosConfSpec: Option[KerberosConfigSpec] = (for {
+      secretName <- existingSecretName
+      secretItemKey <- existingSecretItemKey
+   } yield {
+      KerberosConfigSpec(
+        dtSecret = None,
+        dtSecretName = secretName,
+        dtSecretItemKey = secretItemKey,
+        jobUserName = kubeTokenManager.getCurrentUser.getShortUserName)
+   }).orElse(
+      if (isKerberosEnabled) {
+         Some(HadoopKerberosLogin.buildSpec(
+           conf,
+           kubernetesConf.appResourceNamePrefix,
+           kubeTokenManager))
+       } else {
+        None
+      }
+   )
+
+   override def configurePod(pod: SparkPod): SparkPod = {
+     val hadoopBasedSparkPod = HadoopBootstrapUtil.bootstrapHadoopConfDir(
+        hadoopConfDirSpec.hadoopConfDir,
+        newHadoopConfigMapName,
+        hadoopConfDirSpec.hadoopConfigMapName,
+        pod)
+      kerberosConfSpec.map { hSpec =>
+        HadoopBootstrapUtil.bootstrapKerberosPod(
+          hSpec.dtSecretName,
+          hSpec.dtSecretItemKey,
+          hSpec.jobUserName,
+          krb5File,
+          Some(kubernetesConf.kRBConfigMapName),
+          krb5CMap,
+          hadoopBasedSparkPod)
+      }.getOrElse(
+        HadoopBootstrapUtil.bootstrapSparkUserPod(
+          kubeTokenManager.getCurrentUser.getShortUserName,
+          hadoopBasedSparkPod))
+    }
+
+   override def getAdditionalPodSystemProperties(): Map[String, String] = {
+     val resolvedConfValues = kerberosConfSpec.map { hSpec =>
+       Map(KERBEROS_DT_SECRET_NAME -> hSpec.dtSecretName,
+         KERBEROS_DT_SECRET_KEY -> hSpec.dtSecretItemKey,
+         KERBEROS_SPARK_USER_NAME -> hSpec.jobUserName,
+         KRB5_CONFIG_MAP_NAME -> krb5CMap.getOrElse(kubernetesConf.kRBConfigMapName))
+       }.getOrElse(
+          Map(KERBEROS_SPARK_USER_NAME ->
+            kubeTokenManager.getCurrentUser.getShortUserName))
+       Map(HADOOP_CONFIG_MAP_NAME ->
+         hadoopConfDirSpec.hadoopConfigMapName.getOrElse(
+           kubernetesConf.hadoopConfigMapName)) ++ resolvedConfValues
+    }
+
+   override def getAdditionalKubernetesResources(): Seq[HasMetadata] = {
+     // HADOOP_CONF_DIR ConfigMap
+     val hadoopConfConfigMap = for {
+        hName <- newHadoopConfigMapName
+        hFiles <- hadoopConfigurationFiles
+      } yield {
+       HadoopBootstrapUtil.buildHadoopConfigMap(
+         hName,
+         hFiles)
+      }
+
+     // krb5 ConfigMap
+     val krb5ConfigMap = krb5File.map { fileLocation =>
+       HadoopBootstrapUtil.buildkrb5ConfigMap(
+         kubernetesConf.kRBConfigMapName,
+         fileLocation)
+     }
+
+     // Kerberos DT Secret
+     val kerberosDTSecret = for {
```

Comment:
really, please: `kerberosConfSpec.flatMap(_.dtSecret)`.

Reply:
To my knowledge and after a quick test in the shell, you can't use `flatMap` as this would cause a: `dtSecret is not a member of Nothing` error for one of the input cases:
i.e. `None.flatMap(_.dtSecret)` is the cause of the failure.
 :( I understand that it isn't preferred but, I think, this might be the cleanest way to handle the cases: 
- `None`, 
- `Some(KerberosConfSpec(dtSecret: None,...))`,
- `Some(KerberosConfSpec(dtSecret: Some(Secret),...)`

Reply:
the type of `kerberosConfSpec` is known, no? Then it should work.

```
scala> case class Foo(bar: Option[String])
defined class Foo

scala> 

scala> Some(Foo(Some("bar"))).flatMap(_.bar)
res1: Option[String] = Some(bar)

scala> Some(Foo(None)).flatMap(_.bar)
res2: Option[String] = None

scala> Option[Foo](null).flatMap(_.bar)
res3: Option[String] = None
```

A raw "None" would not work, but you don't have a raw "None" here; you have a `val kerberosConfSpec`, of type `Option[KerberosConfigSpec]`, and value `None`, so you have all the type information you need.